modules/junit-jupiter/src/main/java/org/testcontainers/junit/jupiter/TestcontainersExtension.java

```diff
@@ -0,0 +1,76 @@
+package org.testcontainers.junit.jupiter;
+
+import org.junit.jupiter.api.extension.AfterAllCallback;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.platform.commons.util.Preconditions;
+import org.testcontainers.lifecycle.Startable;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+
+class TestcontainersExtension implements BeforeAllCallback, BeforeEachCallback, AfterAllCallback, AfterEachCallback {
+
+    @Override
+    public void beforeAll(final ExtensionContext context) throws IllegalAccessException {
+        Class<?> testClass = context.getRequiredTestClass();
+        for (final Field field : testClass.getDeclaredFields()) {
+            if (isSharedContainer(field)) {
+                startContainer(testClass, field);
+            }
+        }
+    }
+
+    @Override
+    public void beforeEach(final ExtensionContext context) throws IllegalAccessException {
+        Object testInstance = context.getRequiredTestInstance();
+        for (Field field : testInstance.getClass().getDeclaredFields()) {
+            if (isRestartContainer(field)) {
+                startContainer(testInstance, field);
+            }
+        }
+    }
+
+    @Override
+    public void afterEach(final ExtensionContext context) throws IllegalAccessException {
+        Object testInstance = context.getRequiredTestInstance();
+        for (Field field : testInstance.getClass().getDeclaredFields()) {
```

Comment:
Alternatively, you could store the started containers in the `ExtensionContext` and retrieve them from there. You could even store `CloseableResource` implementations and let Jupiter take care of stopping the containers.